packages/firestore/src/remote/watch_change.ts

```diff
@@ -409,16 +411,81 @@ export class WatchChangeAggregator {
         }
       } else {
         const currentSize = this.getCurrentDocumentCountForTarget(targetId);
+        // Existence filter mismatch. Remove documents to limbo, and raise a
+        // snapshot with `isFromCache:true`.
         if (currentSize !== expectedCount) {
-          // Existence filter mismatch: We reset the mapping and raise a new
-          // snapshot with `isFromCache:true`.
-          this.resetTarget(targetId);
-          this.pendingTargetResets = this.pendingTargetResets.add(targetId);
+          // Apply bloom filter to identify and move removed documents to limbo.
+          const bloomFilterApplied = this.applyBloomFilter(
+            watchChange.existenceFilter,
+            targetId,
+            currentSize
+          );
+          if (!bloomFilterApplied) {
+            // If bloom filter application fails, we reset the mapping and trigger
+            // re-run of the query.
+            this.resetTarget(targetId);
+            this.pendingTargetResets = this.pendingTargetResets.add(targetId);
+          }
         }
       }
     }
   }
 
+  /** Returns wheather a bloom filter removed the deleted documents successfully. */
+  private applyBloomFilter(
+    existenceFilter: ExistenceFilter,
+    targetId: number,
+    currentCount: number
+  ): boolean {
+    const unchangedNames = existenceFilter.unchangedNames;
+    const expectedCount = existenceFilter.count;
+
+    if (!unchangedNames || !unchangedNames.bits) {
+      return false;
+    }
+
+    const {
+      bits: { bitmap = '', padding = 0 },
+      hashCount = 0
+    } = unchangedNames;
+    const normalizedBitmap = normalizeByteString(bitmap).toUint8Array();
+    const bloomFilter = new BloomFilter(normalizedBitmap, padding, hashCount);
```

Comment:
We should guard here against invalid bitmap, padding, and hashCount. If any of them are invalid then this method should return `false`. As it is currently written, it will throw an exception and "crash" Firestore.

To do this, I think the BloomFilter constructor should throw a custom error that we can catch here, log, and return false.

e.g.

```
try {
  bloomFilter = new BloomFilter(normalizedBitmap, padding, hashCount);
} catch (err) {
  if (err instanceof BloomFilterInitializationError) {
    logWarn(..., err);
    return false;
  } else {
    throw err;
  }
}
```

I also assume that `normalizeByteString` can potentially throw an exception if it is given invalid data. We should guard against that as well. I've recommended above to add spec tests to cover this.

Reply:
Maybe throw invalid argument error like this:
throw new FirestoreError(Code.INVALID_ARGUMENT, errorMessage)

Reply:
FirestoreError isn't the right error here because its purpose is to represent errors reported from the backend.